File: AlgorithmDataStructure/LeetCode/25-ReverseList/reverse.c

#include <stdio.h>
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode *next;
};
void dump(struct ListNode *head) {
    if (head == NULL) { printf("list null! \n"); }
    struct ListNode *node = head;
    while (node != NULL) {
        printf("%d -> ", node->val);
        node = node -> next;
    }
    printf("\n");
}

struct ListNode* reverseList(struct ListNode* head){
    if (head == NULL) { return NULL; }
    if (head->next == NULL) { return head; }
    
    struct ListNode *prev = NULL;
    struct ListNode *next = NULL;
    while ( head != NULL ) {
        next = head->next;
        head->next = prev;
        prev = head;
        head = next;
    }
    return prev;
}

struct ListNode* reverseKGroup(struct ListNode* head, int k){
    if (head == NULL) { return NULL; }
    if (head->next == NULL || k <= 1) { return head; }
    
    struct ListNode returnHead = {-1, head};
    struct ListNode *start = &returnHead;
    struct ListNode *end = &returnHead;
    while (end != NULL) {
        for (int i = 0; i < k && end != NULL; i++) {
            end = end->next;
        }
        if (end == NULL) { break; }
        struct ListNode *tempStart = start->next;
        struct ListNode *tempEnd = end->next;
        end->next = NULL;
        //下面这里第一次翻转的时候 改变了 returnHead.next。然后后面 start 指向变了，不会修改 returnHead 了
        start->next = reverseList(tempStart); //这里头部，断开了老的链接，接上翻转后新的 head
        tempStart->next = tempEnd; //这里翻转后的尾部，接上了等待下次翻转的尾部
        end = start = tempStart;
    }
    return returnHead.next;
}


int main(int argc, const char * argv[]) {
    
    struct ListNode l1 = {8, NULL};
    struct ListNode l2 = {4, &l1};
    struct ListNode l3 = {1, &l2};

    struct ListNode l4 = {9, &l3};
    struct ListNode l5 = {6, &l4};
    struct ListNode l6 = {2, &l5};
    dump(&l6);
    
    struct ListNode *l = reverseKGroup(&l6, 3);
    dump(l);
    
    return 0;
}

File: AlgorithmDataStructure/LeetCode/25-ReverseList/reverse.swift

public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init(_ val: Int) {
        self.val = val
        self.next = nil
        
    }
    
    func dump() {
        var l: ListNode? = self
        while l != nil {
            print("\(l!.val)")
            l = l!.next
        }
    }
}

class Solution {
    
    func reverseList(_ head: ListNode?) -> ListNode? {
        if head == nil { return nil }
        if head?.next == nil { return head }
        
        var node: ListNode? = head
        var prev: ListNode?
        var next: ListNode?
        while node != nil {
            next = node?.next
            node?.next = prev
            prev = node
            node = next
        }
        return prev
    }
    
    func reverseKGroup(_ head: ListNode?, _ k: Int) -> ListNode? {
        if head == nil { return nil }
        if head?.next == nil || k <= 1 { return head }
        
        let dummy = ListNode(-1)
        dummy.next = head;
        
        var prev: ListNode? = dummy, end: ListNode? = dummy
        while end != nil {
            for _ in 0 ..< k {
                if end == nil { break; }
                end = end!.next
            }
            if end == nil { break; }
            let start = prev?.next
            let next = end?.next
            end?.next = nil
            prev?.next = reverseList(start)
            start?.next = next
            prev = start
            end = prev
        }
        return dummy.next
    }
    
}

let s = Solution()

var l0 = ListNode(10)
var l1 = ListNode(8)
l1.next = l0
var l2 = ListNode(4)
l2.next = l1
var l3 = ListNode(2)
l3.next = l2
var l4 = ListNode(0)
l4.next = l3

l4.dump()
print("-----")
var r = s.reverseKGroup(l4, 3)
r!.dump()
